Adopt Conversations API

raghubetina opened the pull request:
In Aug 2025, OpenAI added the [Conversations API](https://platform.openai.com/docs/api-reference/conversations). This enables two very nice things for our beginners:

1) Instead of this:

```ruby
chat = AI::Chat.new
previous_response_id = @thread.messages.order(:created_at).last.previous_response_id
chat.previous_response_id = previous_response_id
chat.user(params[:content])
chat.generate!
```

something like this:

```ruby
chat = AI::Chat.new
chat.conversation_id = @thread.conversation_id
chat.user(params[:content])
chat.generate!
```

2) It becomes easy to see reasoning items, tool calls, MCP invocations, etc, [with `GET .../items`](https://platform.openai.com/docs/api-reference/conversations/list-items). So now we can do something like this:

```ruby
chat.verbose # =>

┌────────────────────────────────────────────────────────────────────────────┐
│ Conversation: conv_6903c1eea6cc819695af3a1b1ebf9b390c3db5e8ec021c9a        │
│ Items: 3                                                                   │
└────────────────────────────────────────────────────────────────────────────┘

[
  [0] {
         :id => "msg_0c3db5e8ec021c9a006903c1eefdd88196a6a21a3a27ccb2ac",
    :content => [
      [0] {
        :text => "Search for Ruby programming tutorials and tell me about one",
        :type => :input_text
      }
    ],
       :role => :user,
     :status => :completed,
       :type => :message
  },
  [1] {
        :id => "ws_0c3db5e8ec021c9a006903c1ef56cc8196a1d88c43ba574214",
    :action => {
      :query => "Ruby programming tutorials",
       :type => :search
    },
    :status => :completed,
      :type => :web_search_call
  },
  [2] {
         :id => "msg_0c3db5e8ec021c9a006903c1f001f4819687da6b8895513cfe",
    :content => [
      [0] {
        :annotations => [
          [0] {
              :end_index => 439,
            :start_index => 358,
                  :title => "Learn Ruby | Codecademy",
                   :type => :url_citation,
                    :url => "https://www.codecademy.com/learn/learn-ruby?utm_source=openai"
          },
          [1] {
              :end_index => 764,
            :start_index => 664,
                  :title => "Ruby Programming for Beginners | Udemy",
                   :type => :url_citation,
                    :url => "https://www.udemy.com/course/ruby-programming-for-beginners-course/?utm_source=openai"
          },
          [2] {
              :end_index => 1056,
            :start_index => 965,
                  :title => "Ruby in Twenty Minutes",
                   :type => :url_citation,
                    :url => "https://www.ruby-lang.org/en/documentation/quickstart/?utm_source=openai"
          },
          [3] {
              :end_index => 1289,
            :start_index => 1224,
                  :title => "RubyGuides - Learn Ruby With Awesome Tutorials",
                   :type => :url_citation,
                    :url => "https://www.rubyguides.com/?utm_source=openai"
          },
          [4] {
              :end_index => 1638,
            :start_index => 1541,
                  :title => "Learn to Program (book)",
                   :type => :url_citation,
                    :url => "https://en.wikipedia.org/wiki/Learn_to_Program_%28book%29?utm_source=openai"
          }
        ],
               :text => "If you're looking to learn Ruby programming, here are some tutorials that cater to different learning preferences:\n\n\n\n**Learn Ruby by Codecademy**  \nAn interactive, beginner-friendly course covering fundamental programming concepts, including variables, loops, and object-oriented programming. It offers hands-on exercises and a certificate upon completion. ([codecademy.com](https://www.codecademy.com/learn/learn-ruby?utm_source=openai))\n\n\n\n\n**Ruby Programming for Beginners by Udemy**  \nA comprehensive course designed for beginners, teaching the core fundamentals of Ruby. It includes practical examples and exercises to help you write your own Ruby programs. ([udemy.com](https://www.udemy.com/course/ruby-programming-for-beginners-course/?utm_source=openai))\n\n\n\n\n**Ruby in Twenty Minutes**  \nAn official, concise tutorial that introduces Ruby's syntax and basic concepts. It's perfect for those who prefer a quick, hands-on approach to get started with Ruby. ([ruby-lang.org](https://www.ruby-lang.org/en/documentation/quickstart/?utm_source=openai))\n\n\n\n\n**RubyGuides**  \nA collection of over 150 in-depth guides explaining complex Ruby topics in plain English, suitable for both beginners and experienced developers. ([rubyguides.com](https://www.rubyguides.com/?utm_source=openai))\n\n\n\n\n**'Learn to Program' by Chris Pine**  \nA beginner-level book that introduces programming concepts through Ruby, presented in a structured and engaging manner. It's available in multiple languages and offers a solid foundation for new programmers. ([en.wikipedia.org](https://en.wikipedia.org/wiki/Learn_to_Program_%28book%29?utm_source=openai))\n\n\nEach of these resources offers a unique approach to learning Ruby, so you can choose the one that best fits your learning style and goals. ",
               :type => :output_text,
           :logprobs => []
      }
    ],
       :role => :assistant,
     :status => :completed,
       :type => :message
  }
]
```

Also `chat.items` returns the data in a structured way for e.g. looping through, displaying in ERB, etc.

This patch adds those capabilities to the gem. 

Changes:
- Auto-create conversation on first `generate!` call.
- Store `conversation_id` for subsequent turns.
- Maintain backward compatibility with `previous_response_id`.
- Add `items` for programmatic access to conversation data.
- Add `verbose` for terminal inspection with `amazing_print`.
- Upgrade `openai` gem to 0.34 (Conversations API support).
- Add `amazing_print` as runtime dependency (wanted it anyway to improve how `AI::Chat` prints itself, in a later PR).
- Forking conversations still supported via `previous_response_id`, which takes precedence if both IDs are set (with warning).

Resolves #28

---

To test:

- Run `bundle exec ruby examples/15_conversation_features_comprehensive.rb`
- Mess around with it. Try complicated reasoning, image gen, etc.


<!-- ELLIPSIS_HIDDEN -->


----

> [!IMPORTANT]
> Adopts Conversations API in `AI::Chat`, adding conversation management and inspection features, while maintaining backward compatibility.
> 
>   - **Behavior**:
>     - Auto-create conversation on first `generate!` call in `chat.rb`.
>     - Store `conversation_id` for subsequent turns, maintaining backward compatibility with `previous_response_id`.
>     - Add `items` for programmatic access to conversation data and `verbose` for terminal inspection.
>     - Forking conversations supported via `previous_response_id`, with precedence over `conversation_id` if both are set.
>   - **Dependencies**:
>     - Upgrade `openai` gem to 0.34 for Conversations API support.
>     - Add `amazing_print` as a runtime dependency for improved output formatting.
>   - **Examples**:
>     - Add `15_conversation_features_comprehensive.rb` to demonstrate new conversation features.
> 
> <sup>This description was created by </sup>[<img alt="Ellipsis" src="https://img.shields.io/badge/Ellipsis-blue?color=175173">](https://www.ellipsis.dev?ref=firstdraft%2Fai-chat&utm_source=github&utm_medium=referral)<sup> for 50690598ee4bcd6de954cc72e54a351c876bc731. You can [customize](https://app.ellipsis.dev/firstdraft/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>


<!-- ELLIPSIS_HIDDEN -->

raghubetina committed:
Adopt Conversations API for server-side context

Replace previous_response_id-only approach with OpenAI's
Conversations API. Automatically creates and manages conversations
server-side, enabling full CRUD access to conversation history
including messages, tool calls, reasoning, and web searches.

Changes:
- Auto-create conversation on first generate! call
- Store conversation_id for subsequent turns
- Maintain backward compatibility with previous_response_id
- Add items() for programmatic access to conversation data
- Add verbose() for terminal inspection with amazing_print
- Upgrade openai gem to 0.34 (Conversations API support)
- Add amazing_print as runtime dependency

Forking conversations still supported via previous_response_id,
which takes precedence if both IDs are set (with warning).

raghubetina committed:
Improve verbose output header with box drawing

Replace plain equals signs with Unicode box-drawing characters
for better visual alignment with colorized amazing_print output.

bpurinton commented:
@jelaniwoods @raghubetina, noticing right away that there will be some conflicts with #29. I think that proxy PR is ready to merge though, so probably a rebase will be in order.

raghubetina committed:
Fix spurious forking warning in multi-turn chats

Only set previous_response_id when not using conversations.
Previously both IDs were set after first turn, triggering the
conflict warning on every subsequent call and defeating the
conversation API. Now conversation_id takes precedence naturally
without noise.

bpurinton commented on Gemfile.lock:
```diff
@@ -2,11 +2,14 @@ PATH
+      openai (~> 0.34)
       tty-spinner (~> 0.9.3)
+    openai (0.34.1)
+      connection_pool
```
Can you do a fresh `bundle install` on this branch and see if this diff remains the same? When I bundle on the branch in order to run the examples, I get a diff in the `.lock` file related to the `connection_pool`

raghubetina replied:
I didn't have any diff when I ran `bundle`, but I blew away `Gemfile.lock`, `update`d, and pushed.

raghubetina committed:
Add Contributing section to README

Clarify code style requirements (StandardRB) and other
contribution guidelines for potential contributors.

bpurinton commented on examples/15_conversation_features_comprehensive.rb:
```diff
@@ -0,0 +1,179 @@
+# Feature 8: Forking conversations
+puts "8. Forking Conversations"
+puts "-" * 60
+puts "You can fork a conversation using previous_response_id to create a branch."
```
cool!

raghubetina committed:
Update verbose output example in README

Show Unicode box drawing in documentation to match actual
output after recent formatting improvements.

bpurinton reviewed (approved):
Works well in my testing

raghubetina committed:
Update Gemfile.lock

Refresh dependency versions from bundle operations.

jelaniwoods commented on lib/ai/chat.rb:
```diff
@@ -153,6 +153,28 @@ def last
+      client.conversations.items.list(conversation_id, order: order)
+    end
+
+    def verbose
+      page = items
+
+      box_width = 78
+      inner_width = box_width - 4
+
+      puts
+      puts "┌#{"─" * (box_width - 2)}┐"
+      puts "│ Conversation: #{conversation_id.ljust(inner_width - 14)} │"
+      puts "│ Items: #{page.data.length.to_s.ljust(inner_width - 7)} │"
+      puts "└#{"─" * (box_width - 2)}┘"
+      puts
+
+      ap page.data, limit: 10, indent: 2
+    end
```
When I call `verbose` on a simple conversation I get warned for forking and I don't see the last message in the output.

```rb
chat = AI::Chat.new
chat.user("What is 2 + 2?")
message = chat.generate![:content]
puts "   Message: #{message}"
puts
puts
puts
chat.user("Okay, now summarize the events leading up to the great depression")
chat.generate!
ap chat.messages
puts "\n" * 8
puts "=" * 20
puts
chat.verbose
```

produces

```txt
Both conversation_id and previous_response_id are set. Using previous_response_id for forking. Only set one.
[
    [0] {
            :role => "user",
         :content => "What is 2 + 2?",
        :response => nil
    },
    [1] {
            :role => "assistant",
         :content => "2 + 2 equals 4.",
        :response => {
                      :id => "resp_0f4ad459a7a3b93b006903cf115b6c819091e56c7a357d4129",
                   :model => "gpt-4.1-nano-2025-04-14",
                   :usage => {
                 :input_tokens => 15,
                :output_tokens => 9,
                 :total_tokens => 24
            },
            :total_tokens => 24,
                  :images => []
        },
          :status => :completed
    },
    [2] {
            :role => "user",
         :content => "Okay, now summarize the events leading up to the great depression",
        :response => nil
    },
    [3] {
            :role => "assistant",
         :content => "Certainly! The events leading up to the Great Depression include several economic and financial developments in the 1920s:\n\n1. **Economic Boom of the 1920s**: The decade was marked by rapid economic growth, technological innovation, and widespread optimism. The stock market experienced significant gains, encouraging many to invest heavily.\n\n2. **Speculative Bubble and Stock Market Crash (1929)**: Excessive speculation and buying on margin inflated stock prices beyond their actual value. The stock market crash in October 1929, known as Black Tuesday, triggered a severe decline in stock prices and eroded wealth.\n\n3. **Bank Failures and Financial Panic**: The crash led to a banking crisis, with many banks failing as they couldn't cover depositors' withdrawals, leading to widespread bank closures and loss of savings.\n\n4. **Reduction in Consumer Spending and Investment**: As confidence waned, both consumers and businesses reduced spending and investment, causing industrial production to decline.\n\n5. **Decline in International Trade**: The U.S. and other countries enacted tariffs (like the Smoot-Hawley Tariff Act of 1930), which led to a decline in global trade and worsened economic downturns abroad.\n\n6. **Agricultural Depression**: Farmers faced falling prices and overproduction, adding to economic woes, especially in rural areas.\n\nThese interconnected factors culminated in a severe worldwide economic downturn, known as the Great Depression, which lasted throughout the 1930s and had profound social and economic impacts globally.\n\nWould you like a more detailed explanation of any specific event?",
        :response => {
                      :id => "resp_0f4ad459a7a3b93b006903cf12f6308190964875a2a1fec66c",
                   :model => "gpt-4.1-nano-2025-04-14",
                   :usage => {
                 :input_tokens => 43,
                :output_tokens => 320,
                 :total_tokens => 363
            },
            :total_tokens => 363,
                  :images => []
        },
          :status => :completed
    }
]








====================


┌────────────────────────────────────────────────────────────────────────────┐
│ Conversation: conv_6903cf112ff08190ad1637f78f323e9d0f4ad459a7a3b93b        │
│ Items: 2                                                                   │
└────────────────────────────────────────────────────────────────────────────┘

[
  [0] {
         :id => "msg_0f4ad459a7a3b93b006903cf1175c081909c5a9ddc899f6516",
    :content => [
      [0] {
        :text => "What is 2 + 2?",
        :type => :input_text
      }
    ],
       :role => :user,
     :status => :completed,
       :type => :message
  },
  [1] {
         :id => "msg_0f4ad459a7a3b93b006903cf122ec88190b70c3d8b9d889db5",
    :content => [
      [0] {
        :annotations => [],
               :text => "2 + 2 equals 4.",
               :type => :output_text,
           :logprobs => []
      }
    ],
       :role => :assistant,
     :status => :completed,
       :type => :message
  }
]
```

is this expected?

should conversation_id appear in the `messages` array some where?

raghubetina commented:
@jelaniwoods Could you take over this PR, please? To be merged before or after #29, whichever order you think is better.

jelaniwoods commented:
> jelaniwoods Could you take over this PR, please? To be merged before or after #29, whichever order you think is better.

@raghubetina will do. One clarification question I had:

Currently `conversation_id` and `previous_response_id` are _both_ automatically set and enabled. This causes a the warning message to print and the gem will default to "forking" and using the `previous_response_id`. Is this the intended behavior? Or should `previous_response_id` no longer be set automatically?

raghubetina commented:
> > jelaniwoods Could you take over this PR, please? To be merged before or after #29, whichever order you think is better.
> 
> @raghubetina will do. One clarification question I had:
> 
> Currently `conversation_id` and `previous_response_id` are _both_ automatically set and enabled. This causes a the warning message to print and the gem will default to "forking" and using the `previous_response_id`. Is this the intended behavior? Or should `previous_response_id` no longer be set automatically?

I _think_ I fixed this in https://github.com/firstdraft/ai-chat/pull/31/commits/10ff662803b118f60ebd96639f47ea1b7ea8b08e ?

jelaniwoods commented:
> > > jelaniwoods Could you take over this PR, please? To be merged before or after #29, whichever order you think is better.
> > 
> > 
> > raghubetina will do. One clarification question I had:
> > Currently `conversation_id` and `previous_response_id` are _both_ automatically set and enabled. This causes a the warning message to print and the gem will default to "forking" and using the `previous_response_id`. Is this the intended behavior? Or should `previous_response_id` no longer be set automatically?
> 
> I _think_ I fixed this in [10ff662](https://github.com/firstdraft/ai-chat/commit/10ff662803b118f60ebd96639f47ea1b7ea8b08e) ?

Oh! I didn't have that commit locally. Yes this fixes it. Thank you
